LibGUI/PixelPaint: Introduce a new widget "DynamicWidgetContainer"

Torstennator opened the pull request:
This PR adds a new widget "DynamicWidgetContainer" that allows us to group widgets within a container widget that can be collapsed or detached to a new window. Furthermore if this widget is used as a container for other "DynamicWidgetContainers" it is possible to rearrange the possitions of the widgets within this container. All the modifications that can be done with this widget can  be persisted and reloaded automatically.

This new widget as been introduced to PixelPaint as a playground for the new feature but is easily reusable for other applications.

Details see individual commits.

Preview:

![Peek 2023-11-07 19-03](https://github.com/SerenityOS/serenity/assets/103676658/418e1e39-ec47-47f5-8c66-52ef51b91433)

circl-lastname commented:
Personally the design doesn't fit quite right into the wider Serenity style, I'll leave that up to the maintainers though :^)

diegogangl commented:
Great work @Torstennator :)
IMO this kind of widget is super important, not just for Pixel Paint but for all content editing app (and even Hack Studio). You can see it in Blender, Krita or Kdenlive; there's always going to be an overwhelming amount of tools, options and panels that need to be arranged and managed.

Some notes:
- I'd try to see how it looks without the grey backgound, just the buttons in black.
- Since you are already reserving a row for the buttons, and showing the title when they are minimized. Why not have the title of the child widget next to the buttons all the time? That way it can be removed from the child widget and save some space. Even the outline could be removed and have a single line separating the containers.

cubiclove commented:
Would be awesome if there was a grab handle to rearrange the panels!

Torstennator commented:
Changes since last push:
- buttons have a more decent look now
- it is now possible to rearrange sections and persist that changes (DynamicWidgetContainers within a DynmaicWidgetContainer)

kleinesfilmroellchen commented on Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:
```diff
@@ -0,0 +1,475 @@
+    if (event.button() != MouseButton::Primary)
+        return;
+
+    if (m_expand_collapse_button.contains(event.position()) && !is_container_with_individual_order())
+        set_view_state(view_state() == ViewState::Expanded ? ViewState::Collapsed : ViewState::Expanded);
+
+    if (m_detach_widget_button.contains(event.position()))
+        set_view_state(ViewState::Detached);
```
A more general thing that I was asking myself only after a while of looking at this code: Is there a reason why you're reinventing buttons and labeling and various GUI functionality and layouting calculations for the container controls? It would be much better if you could add a widget that contains the label and the control buttons. That would also allow us to change the primary direction of the container to horizontal instead of vertical, since that only requires to change the widget's layout type to vertical instead of horizontal. I think you can also apply the knurling better in this way.

Torstennator replied:
There was a reason, initially i was aiming to have the buttons drawn as a overlay over the containing widgets and therefore i started to draw them manually as a "on_second_paint" thing. But that did not work out as expected in regards to the mouse events getting catched or propagted to the childs. With that idea the DynamicWidgetContainer would not consume space for itself.
With the current solution the "header" is taking up some space... i'll refactor that to make use of existing button and label widgets.

Torstennator replied:
A new widget that contains the label and buttons has been implemented.

kleinesfilmroellchen commented on Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:
```diff
@@ -0,0 +1,475 @@
+        }
+
+        m_is_dragging = false;
+        set_override_cursor(Gfx::StandardCursor::Arrow);
```
This seems unnecessary and wrong.

Torstennator replied:
If that is not in place, the mouse cursor will only change back if you move the mouse cursor.

kleinesfilmroellchen commented on Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:
```diff
@@ -0,0 +1,475 @@
+
+void DynamicWidgetContainer::doubleclick_event(MouseEvent& event)
+{
+    if (event.button() != MouseButton::Primary || is_container_with_individual_order())
```
Why does a container with individual order not have the double-click functionality? That doesn't make sense from a user perspective.

Torstennator replied:
Iv'e enabled the double click again, there was no real reason for that - i just didn't liked the collapsed main section in pixelpaint - but that should not be a general restriction.

kleinesfilmroellchen commented on Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:
```diff
@@ -0,0 +1,475 @@
+        set_frame_style(m_previous_frame_style.value());
+        m_previous_frame_style = {};
+
+        deferred_invoke([&]() {
+            invalidate_layout();
+        });
```
Pretty sure you can call invalidate_layout directly.

Torstennator replied:
Hm, it is not working as expected without the deferred invoke and resulted in the widgets not always being drawn completely. So I didn't changed the existing solution as that working as expected.

kleinesfilmroellchen commented on Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:
```diff
@@ -0,0 +1,475 @@
+
+void DynamicWidgetContainer::resize_event(ResizeEvent&)
+{
+    constexpr int button_width_and_internal_margin = 24;
+    m_expand_collapse_button.set_x(width() - (button_width_and_internal_margin + content_margins().right()));
+    m_detach_widget_button.set_x(m_expand_collapse_button.x() - 18);
```
Yeah no. Please use a proper child widget for all of these controls.

kleinesfilmroellchen commented on Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:
```diff
@@ -0,0 +1,475 @@
+    });
+
+    VERIFY(destination_widget);
+    VERIFY(source_widget);
```
Instead pass the source widget as a NNRP; the caller must check that it's passing a valid widget.

kleinesfilmroellchen commented on Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:
```diff
@@ -0,0 +1,475 @@
+    if (source_widget == destination_widget)
+        return;
+
+    auto source_pos = children().find_first_index(*source_widget);
+    auto destination_pos = children().find_first_index(*destination_widget);
+
```
No need for abbreviation

kleinesfilmroellchen commented on Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:
```diff
@@ -0,0 +1,475 @@
+    VERIFY(source_pos.has_value());
+    VERIFY(destination_pos.has_value());
+
+    swap(children()[source_pos.value()], children()[destination_pos.value()]);
+
+    // FIXME: Find a better solution to instantly display the new widget order.
+    //        invalidate_layout is not working :/
+    auto childs = child_widgets();
+    for (RefPtr<GUI::Widget> widget : childs) {
+        widget->remove_from_parent();
+        (void)try_add_child(*widget);
+    }
```
I'm pretty sure changes caused by modifying the child list directly are not handled correctly. Can you use the normal APIs instead?

Torstennator replied:
I was trying to improve that by using `remove_child` and `insert_child_before` but that didn't worked out. Somehow it was never working as expected and introduced some weird looking spaces between the moved widgets. For to moment i would like to leave the not perfect but working solution :-/

kleinesfilmroellchen commented on Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:
```diff
@@ -0,0 +1,475 @@
+
+    JsonArray new_widget_order;
+    for_each_child_container([&](auto& child) {
+        (void)new_widget_order.append(child.section_label().ascii_case_insensitive_hash());
+
+        return IterationDecision::Continue;
```
This empty line looks weird

kleinesfilmroellchen commented on Userland/Libraries/LibGUI/DynamicWidgetContainer.h:
```diff
@@ -0,0 +1,104 @@
+        Detached
+    };
+
+    enum class MoveTargetOp {
```
`MoveTargetOperation`

kleinesfilmroellchen commented on Userland/Applications/PixelPaint/PixelPaintWindow.gml:
```diff
@@ -43,59 +43,83 @@
-                layout: @GUI::VerticalBoxLayout {
-                    margins: [6]
+            preferred_height: "grow"
+            detached_size: [200, 640]
```
I don't think this is necessary if we carry over the in-window widget size correctly, as pointed out in an earlier comment.

Torstennator replied:
I've changed the logic so that the new window defaults to the in-window size, but having an alternative other detached size seems quite comfortable as you might want to detach the widget because it should display more information when more space is available - so you don't need to resize the window manually after detaching...

kleinesfilmroellchen commented on Userland/Applications/PixelPaint/main.cpp:
```diff
@@ -53,8 +54,10 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
     window->on_close_request = [&]() -> GUI::Window::CloseRequestDecision {
-        if (main_widget->request_close())
+        if (main_widget->request_close()) {
+            GUI::DynamicWidgetContainer::close_all_detached_windows();
```
This should not be necessary. Please figure out how to call that function at application exit; or better yet, for each separate container widget when it's destructed (RAII).

Torstennator replied:
For the moment i was not able to find a reliable working solution e.g. closing all windows within the destructor. That destructor somehow never was called when the widget has been cunstructed via a gml file - I have no idea whats the reason behind that.
For the moment i close all the open windows in the destructor as when the widget was not constructed via gml and the old solution where they are closed via the static function.

kleinesfilmroellchen commented on Userland/Applications/PixelPaint/HistogramWidget.cpp:
The last commit can be its own pull request, since it's a separate optimization. Since I think the rest of this PR will require a lot more work, this would help with getting the fix in sooner.

Torstennator replied:
As this is currently not a problem, because the widget is not changing size, i would keep it with this pr as it becomes a problem with the new dynamic behavior of this pr.

kleinesfilmroellchen commented on Userland/Applications/PixelPaint/HistogramWidget.cpp:
```diff
@@ -48,49 +48,41 @@ ErrorOr<void> HistogramWidget::rebuild_histogram_data()
     painter.add_clip_rect(event.rect());
     int bottom_line = height() - 1;
     float step_width = static_cast<float>(width()) / 256;
+    auto calc_y_with_scaled_value = [this, bottom_line](int color_count, int max_value_count) {
```
I can't grok that lambda's name, please find something better.

kleinesfilmroellchen reviewed (changes requested):
While this feature is very cool, it needs a lot more work for it to be Serenity-ready. My main points are:
- Use more non-nullable containers; please also check things that I didn't specifically point out. This reduces footguns and clarifies intent to the reader.
- Use the power of LibGUI for the container controls instead of rewriting your own primitive button controls. There's many problems with the current controls (magic numbers, bad icons, no interactivity, no keyboard motion or keyboard shortcuts possible, bad event handling, no possibility for a horizontal container mode, no respect for the knurling on/off setting) and all of those would solve themselves instantly if you just used a separate subwidget. I'd encourage you to write it in GML for simplicity; it will of course require another separate widget but the result will be much much better.
- Do not magic number any dimensions, paddings, margins, ... unless you need to account for off-by-one in graphics. We have a huge amount of magic number dimensions in the system for historical reasons, but it's nothing we should do in new code.

Torstennator commented:
@kleinesfilmroellchen Thanks for the detailed review!
I'll go through all your points as i find the time to do so, just started with the first steps...

Torstennator commented on Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:
```diff
@@ -0,0 +1,475 @@
+        painter.fill_rect_with_gradient(Gfx::IntRect(0, height() - 1, width(), height()), palette().threed_shadow1().lightened(), palette().button());
+    }
+
+    auto controls_color = m_hover_over_control_indicator ? palette().threed_shadow2() : palette().hover_highlight().lightened(0.7f);
```
That was a mistake, as i just recognized by now that the color i was looking for is the rubber band...

Torstennator commented on Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:
```diff
@@ -0,0 +1,475 @@
+            for (RefPtr<GUI::Widget> child : new_child_order)
+                (void)try_add_child(*child);
+        } else {
+            int persisted_state = Config::read_i32(config_domain(), "DynamicWidgetContainers"sv, section_label(), (int)ViewState::Expanded);
+            set_view_state(static_cast<ViewState>(persisted_state));
```
Currently my logic is to persist the view state or the order of child widgets whereas to order is also some kind of view state. So i use one or the other based on `is_container_with_individual_order()`.

Torstennator commented on Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:
```diff
@@ -0,0 +1,475 @@
+
+    m_controls_spacer = &add<GUI::Label>();
+    // Set the width to zero so that our widget can receive the click events.
+    m_controls_spacer->set_fixed_size({ 0, 12 });
```
Since there is now a new widget that has all the buttons and the label - this is now obsolete.

Torstennator commented on Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:
```diff
@@ -0,0 +1,475 @@
+        painter.set_pixel(x + 1, y + 1, palette().threed_highlight());
+    };
+    for (int i = 0; i < 5; ++i)
+        paint_knurl(1, 0 + (i * 3));
```
For the moment i removed the knurl stuff - not sure if i liked it.

Torstennator commented on Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:
```diff
@@ -0,0 +1,475 @@
+                if (widget.ptr() == this->m_controls_spacer)
+                    continue;
+                widget->remove_from_parent();
+                widget->set_visible(true);
```
Was thinking about it, where is already solved that on pixel paint.... that kind of widgets would need to be wrapped in another widget, so that the visibility does not get screwed up by the container.

Torstennator commented on Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:
```diff
@@ -0,0 +1,475 @@
+    if (m_detached_widgets_window.is_null()) {
+        m_detached_widgets_window = TRY(GUI::Window::try_create());
+        m_detached_widgets_window->set_title(section_label().to_deprecated_string());
+        m_detached_widgets_window->set_window_type(WindowType::Normal);
```
For the moment i really want the have the window independent from the main window and not blocking the main window while it's active.
A modal window that e.g. blocks the main window could probably be some future extension if that is needed.

Torstennator commented on Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:
```diff
@@ -0,0 +1,475 @@
+        root_container->set_layout<GUI::VerticalBoxLayout>();
+        root_container->set_frame_style(Gfx::FrameStyle::Window);
+        auto transfer_childs = [this](auto reciever, auto childs) {
+            for (RefPtr<GUI::Widget> widget : childs) {
```
I was trying to change that, but the compiler was not happy about the const so i've not changed that as suggested.

kleinesfilmroellchen commented:
Excellent progress! On my phone so a full review is hard, but a few points to consider:
- Optional<NNRP> can just be RP and Optional<RP> also is just a RP.
- Use static GML generation, dynamic GML generation is for Playground only.
- The labels and buttons shouldn't need a fixed height since they can correctly calculate their min height. (They are already "shrink")
- Check spelling and grammar in the new docs.

Torstennator commented:
Hey @kleinesfilmroellchen ,
I've addressed your latest remarks and hope to get the pr closer to a state where i can remove the "Draft" state. Let me know if you have any further concerns.

circl-lastname commented:
Bump, this is too cool

AtkinsSJ commented on Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:
```diff
@@ -0,0 +1,464 @@
+        auto detached_window = TRY(GUI::Window::try_create());
+        detached_window->set_title(section_label().to_deprecated_string());
+        detached_window->set_window_type(WindowType::Normal);
+        if (m_detached_size.has_value())
+            detached_window->resize(detached_size());
+        else
+            detached_window->resize(size());
```
Mixing and matching `m_detached_size` and `detached_size()` feels awkward. Maybe `detached_size()` should just return an Optional size always?

Torstennator replied:
That is a bit clunky because it is used as a size property and therfore does not support the optional :/ I've introduced a `has_detached_size()` function.

AtkinsSJ commented on Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:
```diff
@@ -0,0 +1,464 @@
+
+        detached_window->on_close = [this, root_container, transfer_children]() {
+            transfer_children(this, root_container->child_widgets());
+            this->set_view_state(ViewState::Expanded);
+            unregister_open_window(m_detached_widgets_window.value());
+            this->m_detached_widgets_window = {};
```
Avoid `this->` when it's not needed.

AtkinsSJ commented on Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:
```diff
@@ -0,0 +1,464 @@
+        root_container->set_layout<GUI::VerticalBoxLayout>();
+        root_container->set_frame_style(Gfx::FrameStyle::Window);
+        auto transfer_children = [this](auto reciever, auto childs) {
+            for (NonnullRefPtr<GUI::Widget> widget : childs) {
```
`auto& widget: childs` should be fine and avoid copying the refptrs.

AtkinsSJ commented on Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:
```diff
@@ -0,0 +1,464 @@
+        root_container->set_frame_style(Gfx::FrameStyle::Window);
+        auto transfer_children = [this](auto reciever, auto childs) {
+            for (NonnullRefPtr<GUI::Widget> widget : childs) {
+                if (widget == this->m_controls_widget)
```
`this->` isn't necessary.

AtkinsSJ commented on Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:
```diff
@@ -0,0 +1,464 @@
+        m_is_dragging = false;
+
+        auto label_widget = m_controls_widget->find_descendant_of_type_named<GUI::LabelWithEventDispatcher>("section_label");
+        // Change the cursor back to normal after dragging is finished. Otherwise the curse will only change if the mouse moves.
```
"curse" -> "cursor"

AtkinsSJ commented on Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:
```diff
@@ -0,0 +1,464 @@
+        update();
+    }
+
+    auto label_widget = m_controls_widget->find_descendant_of_type_named<GUI::LabelWithEventDispatcher>("section_label");
+    label_widget->update_cursor(changed_cursor);
```
`DynamicWidgetContainerControls::get_event_dispatcher()`? Though, maybe that method should have a better name. (`get_section_label()`?)

It also feels a little weird to be reaching inside and modifying its child widgets instead of telling it to do so. But no big deal.

AtkinsSJ commented on Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:
```diff
@@ -0,0 +1,464 @@
+void DynamicWidgetContainer::second_paint_event(PaintEvent&)
+{
+    GUI::Painter painter(*this);
+    painter.fill_rect_with_gradient(Gfx::IntRect(0, height() - 1, width(), height()), palette().threed_shadow1().lightened(), palette().button());
```
What's the reason for making this a gradient? It seems a little arbitrary.

Torstennator replied:
There is no reason, that was just how i would like to draw a small separation between the widgets . I've changed that to a solid line.

AtkinsSJ commented on Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:
```diff
@@ -0,0 +1,464 @@
+        return;
+
+    if (m_is_dragging) {
+        // FIXME: Would be nice if we could paint outside our own boundaries.
```
The reason you can't is that the `painter(*this)` constructor above is clipping to this widget's bounds. I think `painter.clear_clip_rect()` would solve this but I haven't tested it myself. Might take some fiddling with.

AtkinsSJ commented on Userland/Libraries/LibGUI/DynamicWidgetContainer.h:
```diff
@@ -0,0 +1,96 @@
+    bool show_controls() const { return m_show_controls; }
+    void set_parent_container(RefPtr<GUI::DynamicWidgetContainer>);
+    bool check_has_move_target(Gfx::IntPoint, MoveTargetOperation);
+    // FIXME: this should not be a public api and static - but currently the destructor is not being called when the widget was created via a gml file.
```
:thinking: Is this true for all widgets created from GML? If you can find a reduced test case it would be useful to create an issue about this. (Even if you can't reduce it, create an issue with whatever you know.)

cubiclove commented on Userland/Applications/PixelPaint/PixelPaintWindow.gml:
```diff
@@ -43,59 +43,80 @@
-        @GUI::Widget {
-            fill_with_background_color: true
+        @GUI::DynamicWidgetContainer {
+            section_label: "Tools"
```
Would it be better to rename this to 'Editor Panels', 'Studio Panels' or something?
Currently the box on the left with all of the brushes and selection tools is called the Tool Panel, so it might be a little confusing (in terms of documentation).

Torstennator commented on Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp:
```diff
@@ -0,0 +1,464 @@
+            }
+
+            Vector<NonnullRefPtr<Widget>> new_child_order;
+            auto child_container = child_containers();
```
As the compiler was not happy with the name `child_containers` due to interference with the function name an auto deduction i changed the name to `containers`.

Torstennator commented:
> This is very, very cool and something we really need! :^)
> 
> Most of this feedback is superficial things, generally this seems good to me.
> 
> I have noticed oddness when playing around with this though. When launching Pixel Paint, some tool windows will appear slightly too small and without their widgets being visible. Collapsing/expanding them, or detaching them, doesn't help. For example, I have this with the "Tool Properties" widget. I have to switch tool to make its contents appear. It's the same if I show the histogram widget, it's too small and doesn't display the histogram.
> 
> Also, there's mention of knurling in the code, but I don't see any? Knurling is specifically the visual stripes to show something is resizable or moveable.

I've double checked the issue with the "Tool Properties" and the same behaviour is currently on the master branch. It only shows the properties if you change the tool at least once - so that seems to be nothing new but should be fixed to display the properties for the initial selected tool without the need to change the tool.
Regarding the histogram widget - i was not able to reproduce that - but with this pr i also changed the logic for the histogram widget so that it only shows something if the image has some data to display.
The knureling was used while developing this widget but i got rid of rendering them while making review changes - now the m_move_widget_knurl is only used as the region where the cursor changes to a "Hand" to indicate that it can be moved.

Torstennator commented:
Hi @AtkinsSJ, I'm not sure if there is still a todo on my side. I don't see any open topics for this pr but the status tells me that there is still a requested change from you.
Please let me know if there is still some open stuff...
Greetings
